staging/src/k8s.io/apiserver/pkg/storage/etcd3/sizer.go

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2025 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcd3
+
+import (
+	"sync"
+
+	"go.etcd.io/etcd/api/v3/mvccpb"
+)
+
+func newSizeCache() *sizeCache {
+	return &sizeCache{
+		perKey: make(map[string]sizeRevision),
+	}
+}
+
+// sizeCache efficiently estimates the average object size based on the last observed state of individual keys.
+// By plugging sizeCache into GetList and Watch functions, a fairly accurate estimate of object sizes can be maintained
+// without additional requests to the underlying storage.
+// To handle potential out-of-order or incomplete data, it uses a per-key revision to identify the newer state.
+// This approach may lead to key leakage if delete events are not observed, thus AverageObjectSize function should be called periodically with the current list of keys.
```

Comment:
This is tricky, but IIUC we will be running it from higher layer via the same mechanism that we use for metrics, right?
I would extend this comment that this is how it will be used in apiserver.

Reply:
> This is tricky, but IIUC we will be running it from higher layer via the same mechanism that we use for metrics, right?

Thinking now that we should not depend on external caller. I think there is a configuration that would not make any calls. For example if APF is disabled.

Reply:
What do you think about running a background goroutine to ensure keys are cleaned? 

PoC https://github.com/serathius/kubernetes/commit/e99597dbfa0cdd3e2ba4302bf0b775c66f2e82d6 the only problem is that store is currently stateless, so it doesn't have a way to for expose `Close` function.

Reply:
Didn't look into details yet, but at the high-level I prefer that option much more.